code/datums/components/cleave_attack.dm

```diff
@@ -0,0 +1,141 @@
+	if(swing_speed_mod)
+		src.swing_speed_mod = swing_speed_mod
+	if(requires_wielded)
+		src.requires_wielded = requires_wielded
+	if(no_multi_hit)
+		src.no_multi_hit = no_multi_hit
+	if(cleave_end_callback)
+		src.cleave_end_callback = cleave_end_callback
+	set_cleave_effect(cleave_effect)
+
+/// Sets the cleave effect to the specified effect, or based on arc size if one wasn't specified.
+/datum/component/cleave_attack/proc/set_cleave_effect(new_effect)
+	if(new_effect)
+		cleave_effect = new_effect
+		return
+	switch(arc_size)
+		if(0 to 120)
+			cleave_effect = /obj/effect/temp_visual/dir_setting/firing_effect/sweep_attack
+		if(120 to 240)
+			cleave_effect = /obj/effect/temp_visual/dir_setting/firing_effect/sweep_attack/semicircle
+		else
+			cleave_effect = /obj/effect/temp_visual/dir_setting/firing_effect/sweep_attack/full_circle
+
+/datum/component/cleave_attack/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_ITEM_AFTERATTACK, PROC_REF(on_afterattack))
+
+/datum/component/cleave_attack/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_ATOM_EXAMINE, COMSIG_ITEM_AFTERATTACK))
+
+/datum/component/cleave_attack/proc/on_examine(atom/examined_item, mob/user, list/examine_list)
+	var/arc_desc
+	switch(arc_size)
+		if(0 to 90)
+			arc_desc = "narrow arc"
+		if(90 to 180)
+			arc_desc = "wide arc"
+		if(180 to 270)
+			arc_desc = "very wide arc"
+		if(270 to INFINITY)
+			arc_desc = "full circle"
+	examine_list += "It can swing in a [arc_desc]."
+
+/datum/component/cleave_attack/proc/on_afterattack(obj/item/item, atom/target, mob/user, proximity_flag, click_parameters)
+	if(proximity_flag || user.a_intent != INTENT_HARM)
+		return // don't sweep on precise hits or non-harmful intents
+	perform_sweep(item, target, user, click_parameters)
+
+/datum/component/cleave_attack/proc/perform_sweep(obj/item/item, atom/target, mob/living/user, params)
+	if(user.next_move > world.time)
+		return // don't spam it
+	if(requires_wielded && !HAS_TRAIT(item, TRAIT_WIELDED))
+		return // if it needs to be wielded, check to make sure it is
+
+	// some information we're going to need later
+	var/turf/user_turf = get_turf(user)
+	var/turf/center_turf = get_turf_in_angle(get_angle(user, target), user_turf)
+	var/facing_dir = get_dir(user, center_turf)
+	var/swing_direction = (user.active_hand_index % 2) ? -1 : 1
+
+	// make a list of turfs to swing across
+	var/list/turf_list = list()
+	var/turfs_count = round(arc_size / 90, 1)
+	for(var/i in -min(turfs_count, 3) to min(turfs_count, 4)) // do NOT hit the same tile more than once
+		turf_list.Add(get_step(user_turf, turn(facing_dir, i * 45 * swing_direction)))
+
+	// do some effects so everyone knows you're swinging a weapon
+	playsound(item, 'sound/weapons/punchmiss.ogg', 50, TRUE)
+	new cleave_effect(user_turf, facing_dir)
+
+	// now swing across those turfs
+	ADD_TRAIT(item, TRAIT_CLEAVING, REF(src))
+	attack_loop:
```

Comment:
man what the hell is this please use human coding styles like while or something

Reply:
This is because it's a for loop within a for loop, and it needs to break the outer loop instead of the inner loop in order to stop the whole thing or it will continue on to the next tile. This was the simplest solution to it that I could find, but if you want me to do it another way then sure.

Reply:
Instead of break make a proc to handle the the end (removing trait, invoking callback, etc) and call that then early return